File: backend/src/api/routers/sources.py
from fastapi import APIRouter

from api.dependencies import Authenticated, Collections, Sources
from api.responses import READ_ERROR_CODES, WRITE_ERROR_CODES, error_responses
from sources.schemas import QueryPreview, QueryPreviewInput, SourceView

router = APIRouter(prefix="/sources", tags=["sources"])


@router.get(
    "",
    response_model=list[SourceView],
    responses=error_responses(*READ_ERROR_CODES),
    operation_id="listSources",
)
def source_catalog(
    service: Sources,
    collections: Collections,
    owner: Authenticated,
) -> list[SourceView]:
    return service.catalog(collections.source_runtime_statuses())


@router.post(
    "/query-preview",
    response_model=QueryPreview,
    responses=error_responses(*WRITE_ERROR_CODES),
    operation_id="previewSourceQueries",
)
def query_preview(data: QueryPreviewInput, service: Sources, owner: Authenticated) -> QueryPreview:
    return service.preview(data)

File: frontend/src/api/sources.ts
// @ts-ignore
/* eslint-disable */
import { request, type RequestOptions } from "../request";

/** Source Catalog GET /api/sources */
export async function listSources(options?: RequestOptions) {
  return request<API.SourceView[]>("/api/sources", {
    method: "GET",
    ...(options || {}),
  });
}

/** Query Preview POST /api/sources/query-preview */
export async function previewSourceQueries(
  body: API.QueryPreviewInput,
  options?: RequestOptions
) {
  return request<API.QueryPreview>("/api/sources/query-preview", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    data: body,
    ...(options || {}),
  });
}
